boa_engine/src/optimizer/mod.rs

```diff
@@ -0,0 +1,115 @@
+use crate::Context;
+use boa_ast::{visitor::VisitorMut, Expression};
+use std::{fmt, ops::ControlFlow};
+
+/// The action to be performed after an optimization step.
+#[derive(Debug)]
+
+pub(crate) enum PassAction<T> {
+    /// Keep the node, do nothing.
+    Keep,
+
+    /// The node was modified inplace.
+    Modified,
+
+    /// Replace the node.
+    Replace(T),
+}
+
+/// Contains statistics about the optimizer execution.
+#[derive(Debug, Default, Clone, Copy)]
+pub struct OptimizerStatistics {
+    /// How many times was the optimization run in total.
+    pub constant_folding_run_count: usize,
+
+    /// How many passes did the optimization run in total.
+    pub constant_folding_pass_count: usize,
+}
+
+impl fmt::Display for OptimizerStatistics {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        writeln!(f, "Optimizer {{")?;
+        writeln!(
+            f,
+            "    constant folding: {} run(s), {} pass(es) ({} mutating, {} checking)",
+            self.constant_folding_run_count,
+            self.constant_folding_pass_count,
+            self.constant_folding_pass_count
+                .saturating_sub(self.constant_folding_run_count),
+            self.constant_folding_run_count
+        )?;
+        writeln!(f, "}}")?;
+        Ok(())
+    }
+}
+
+/// This represents an AST optimizer.
+#[derive(Debug)]
+pub(crate) struct Optimizer<'context, 'host> {
+    statistics: OptimizerStatistics,
+    context: &'context mut Context<'host>,
+}
+
+impl<'context, 'host> Optimizer<'context, 'host> {
+    /// Create a optimizer.
+    pub(crate) fn new(context: &'context mut Context<'host>) -> Self {
+        Self {
+            statistics: OptimizerStatistics::default(),
+            context,
+        }
+    }
+
+    /// Run the constant folding optimization on an expression.
+    fn run_constant_folding_pass(&mut self, expr: &mut Expression) -> bool {
+        self.statistics.constant_folding_run_count += 1;
+
+        let mut has_changes = false;
+        loop {
+            self.statistics.constant_folding_pass_count += 1;
+            let mut walker = Walker::new(|expr| -> PassAction<Expression> {
+                ConstantFolding::fold_expression(expr, self.context)
+            });
+            // NOTE: postoder traversal is optimal for constant folding,
+            // since it evaluates the tree bottom-up.
+            walker.walk_expression_postorder(expr);
+            if !walker.changed() {
+                break;
+            }
+            has_changes = true;
+        }
+        has_changes
+    }
+
+    fn run_all(&mut self, expr: &mut Expression) {
+        // TODO: Add more optimizations here...
+        // loop {
+        //     let mut changed = false;
+        //     changed |= self.run_constant_folding_pass(expr);
+        //
+        //     if !changed {
+        //         break;
+        //     }
+        // }
+        self.run_constant_folding_pass(expr);
```

Comment:
About the optimizer pass architecture, maybe we could copy `dfdx`'s design of [`Module`](https://docs.rs/dfdx/latest/dfdx/nn/trait.Module.html), where a trait is implemented for the `(P1, P2, P3, P4)` tuple with `P1/P2/P3/P4: OptimizationPass`. That way, if we wanted to add a new optimization pass `Pass4` we just implement `OptimizationPass` for it, then we'd just need to change something like `(pass1, pass2, pass3).run(expr)` into `(pass1, pass2, pass3, pass4).run(expr)`.

Reply:
Hmmm. interesting! Will look into it! :)

Reply:
For this PR since it's only one optimization (seems unnecessary), I'll leave it as is for now, when we implement other optimizations we can see how well this architecture fits :)